### ISSUE TITLE
CLI: kontena service logs --tail skips loglines

### ISSUE THREAD

**Issue by matti** [kontena/kontena#959]
```
$ kontena service logs --tail sidekiq | grep QUEUE
```

then run

```
$ kontena container exec railsapp-1 rails runner \"puts EnqueueJobForAllRecordsOfClass.perform_async(\'SomeModule::SomeClass\', \'User\')\"
```

You are expecting to see log lines with numbers 1-600 in log tailing, but you only see a random batch from the middle of the log lines.

The following command shows the correct log lines:

```
$ kontena service logs --lines=10000 sidekiq | grep QUEUE
```

**Comment by SpComb** [comment]
I'm unable to reproduce this on 0.15.1 or current master. I'm testing with a minimal docker `test` image:

```
FROM debian:jessie

CMD ["bash", "-c", "sleep 10 && for i in {1..600}; do echo 'log' $i; done; sleep 60"]
```

Running `kontena service create test test`, I haven't been able to spot any missed log lines in `kontena service logs --tail test`.

How exactly does the `kontena service logs` command behave in this case with the current release? Does it exit (see #978) or start eating memory (see #972)?

Can you test if this still happens with the current kontena master or upcoming 0.15.2 bugfix release containing those fixes?

**Comment by matti** [comment]
@juhazi can you jump in here?

**Comment by SpComb** [comment]
Second unsuccesful attempt to repro with the "exec in client to trigger server logs" flow, using the minimal `test-server` and `test-client` Docker images here: https://gist.github.com/SpComb/559ab58a232d679183fdaddb781ba9dc

`kontena service create test-server test-server && kontena service start test-server`
`kontena service create test-client test-client && kontena service start test-client`

And then watching: `kontena service logs -t test-server` while running `kontena container exec test-client-1 /test.sh`. All of the `/test/1` to `/test/600` logs show. 

I suspect this might be something related to the multiple-partial-json-chunks issue fixed in #972 combined with some SocketError that triggers a CLI-side retry of the `get_stream("services/..../container_logs")`request, but that's tricky to confirm. Alternative is some bug in the server-side stream-loop mongodb queries.

**Comment by juhazi** [comment]
confirmed that your repro is also unsuccessful in our grid.

also our original repro no longer reproducres either. Kontena in our grid has been updated to 0.15.1 after the original issue report.

I did not make a note of kontena version when originally reporting this. :(

I recommend closing this.

**Comment by SpComb** [comment]
BTW, are you sure this was with `kontena service logs` and not `kontena app logs`?

I just submitted some fixes in #984 that could easily have caused this kind of issue with `kontena app logs`...

**Comment by juhazi** [comment]
it's possible. I mix and match `kontena grid logs`, `kontena service logs` and `kontena app logs` as needed.

**Comment by SpComb** [comment]
With the fixes for `kontena app logs` (#987) and `kontena service logs` (#972) in master, I hope this should be fixed.

Please re-open if this still happens with git master or the upcoming 0.15.2 release.
